File: 1275-validate-binary-tree-nodes/validate-binary-tree-nodes.py
class Solution:

    def findParent(self,x):
        if self.parentArr[x]==x:
            return x
        self.parentArr[x]=self.findParent(self.parentArr[x])
        return self.parentArr[x]

    def Union(self,parent,child):
        child_ka_parent = self.findParent(child)
        if child_ka_parent != child:
            return False
        parent_ka_parent = self.findParent(parent)
        if parent_ka_parent == child:
            return False
        self.parentArr[child]=parent
        self.components-=1
        return True            
    def validateBinaryTreeNodes(self, n: int, leftChild: List[int], rightChild: List[int]) -> bool:
        self.components = n
        self.parentArr = [i for i in range(n)]

        for i in range(n):
            parent = i
            left_child = leftChild[i]
            right_child = rightChild[i]
            if left_child!=-1 and self.Union(parent,left_child)==False:
                return False

            if right_child != -1 and self.Union(parent,right_child)==False:
                return False
        
        return self.components == 1

        
        

File: 1275-validate-binary-tree-nodes/validate-binary-tree-nodes.cpp
class Solution {
public:
    vector<int>parentArr;
    int components;
    int findParent(int x) {
        if (parentArr[x]==x) return x;
        return parentArr[x]=findParent(parentArr[x]);
    }
    bool Union(int parent,int child){
        int child_ka_parent = findParent(child);
        if(child_ka_parent != child) return false;
        int parent_ka_parent = findParent(parent);
        if (parent_ka_parent == child) return false;

        parentArr[child]=parent;
        components--;
        return true;
    }

    bool validateBinaryTreeNodes(int n, vector<int>& leftChild, vector<int>& rightChild) {
        components = n;
        parentArr.resize(n);
        iota(begin(parentArr),end(parentArr),0);

        for(int i=0;i<n;i++){
            int parent = i;
            int left_child = leftChild[i];
            int right_child = rightChild[i];
            if(left_child!=-1 && !Union(parent,left_child)) return false;
            if (right_child!=-1 && !Union(parent,right_child)) return false;
        }

        return components == 1;

    }
};